Configurable git sync command (and other git args)

cognifloyd opened the pull request:
I want to make sure that no merge commits are ever added to my upstream tracking branch. So, I need to use one of these:

- `git pull --ff-only <commitish>`
- `git merge --ff-only <commitish>`
- `git reset --hard <commitish>`

To make this possible, add a new `git_sync_command` input option. This defaults to `pull` to retain the current behavior of using `git pull <commitish>`.

This also makes args for `git checkout`, `git fetch`, `git log`, and `git push` configurable.

~This is built on top of #5, so #5 should be merged first.~ _merged_

aormsby commented:
Yeah, I can imagine a lot of people needing something like this. For now, I'll merge this in (with #5 before it) and add a new version tag (1.2).

Beyond this PR, I think I'll add similar options for all the other git commands where it makes sense and bump the whole thing to 2.0. This kind of flexibility will probably be nice to have in the long run. I'll make a note in that release if I change the var name or anything like that (I probably won't).

Thanks for all the input, man. I'm glad you're getting some use out of this. 👍

aormsby commented:
Oh, one small thing - let's make `required: false` so it's not actually necessary to include the setting in a workflow if people just want it to default to `pull`.

cognifloyd commented:
I don't see any other git commands where configuring the command adds obvious value. /shrug

I rebased and adjusted `git_sync_command` to be `required: false`. I also removed a note that said `target_branch` is optional (it is required in action.yaml).

cognifloyd commented:
Ah. someone might want `--recurse-submodules`:
- `git checkout --recurse-submodules <target_branch>`
- `git fetch --recurse-submodules upstream <upstream_branch>`

And they might want to customize the log output:
- `git log upatream/<upstream_branch> <commitsh>..HEAD <log_format_args>`

I don't see any options in the man page that look promising for `git remote add`.

I'll add those as well.

cognifloyd commented:
So, I added all of the git arg vars where I think it might make sense. Do you want me to split this into 2 PRs?

cognifloyd commented:
I debated renaming `GIT_SYNC_METHOD` to `GIT_SYNC_COMMAND`. Any other ideas on how to name these?

cognifloyd commented:
Rebased and re-organized commits to have a sane history.
If you wnat the other commands to be in a separate PR, then I can pull out the "Make other git commands configurable" commit.

aormsby commented:
Oh cool, thanks for that. 🙂 

> Ah. someone might want --recurse-submodules

Yes, same thought here. I realized I was missing that and perhaps `--tags` related options, but your new var should cover those needs well.

> I also removed a note that said target_branch is optional (it is required in action.yaml).

I'm actually looking at my yaml and wondering why I set `required: true` to inputs that also have default values... It seems odd to me now in some cases, but I can revisit that choice later and remember what I was thinking. For now, this is okay.

> Do you want me to split this into 2 PRs?

No, just the one is okay. I can look over it all again in the morning when my mind is a little fresher.

aormsby commented on upstream-sync.sh:
```diff
@@ -36,15 +38,16 @@ fi
+# sync from upstream to target_branch
+echo 'Syncing...' 1>&1
+# sync_command examples: "pull", "merge --ff-only", "reset --hard"
+git ${INPUT_GIT_SYNC_COMMAND} "${INPUT_UPSTREAM_BRANCH}"
```
Oops, you removed 'upstream' here. Please re-insert that between the expanded vars so the git command works with the expected remote location.

cognifloyd replied:
good catch

aormsby commented on action.yaml:
```diff
@@ -25,6 +25,31 @@ inputs:
     required: true
     default: 'master'
 
+  git_checkout_extra_args:
+    description: 'Any extra args to pass to `git checkout` like --recurse-submodules (only used when current branch is not target_branch), default = ""'
+    required: false
+    default: ''
+
+  git_fetch_extra_args:
+    description: 'Any extra args to pass to `git fetch` like --recurse-submodules, default = ""'
+    required: false
+    default: ''
```
Also, I'd like you to remove 'extra' from the var names to slim them down a little. Less to type, fewer potential errors.

cognifloyd replied:
done

aormsby commented on action.yaml:
```diff
@@ -25,6 +25,31 @@ inputs:
+    required: false
+    default: 'pull'
+
+  git_push_extra_args:
```
Same here with 'extra'

cognifloyd replied:
done

aormsby commented:
thanks, included in release 2.0
